crates/air-lib/interpreter-data/src/executed_state.rs

```diff
@@ -53,19 +57,67 @@ pub enum CallResult {
     /// A corresponding call's been already executed with such value as a result.
     Executed(ValueRef),
 
-    /// call_service ended with a service error.
-    #[serde(rename = "failed")]
-    CallServiceFailed(i32, Rc<String>),
+    /// The call returned a service error.
+    ///
+    /// The `JValue` has to be a two element array `[i32, String]`.
+    Failed(Rc<CID<ServiceResultAggregate>>),
 }
 
+/*
+ * Previous data scheme was (Scalar as an example, other cases are similar):
+ *
+ * Scalar(CID<JValue>) ---<CID>----> JValue
+ *
+ * New data scheme is much more sofisticated:
+ *
+ * ```
+ * Scalar(CID<ServiceResultAggregate>) ---+
+ *                                        |
+ *             +--------------------------+
+ *             |
+ *             +---<CID>----> ServiceResultAggregate:
+ *                                value_cid -------------<CID>----> JValue
+ *                                tetraplet_cid ---------<CID>----> SecurityTetraplet
+ *                                argument_hash
+ * ```
+ */
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 #[serde(rename_all = "snake_case")]
 pub enum ValueRef {
-    Scalar(Rc<CID<JValue>>),
+    /// The call value is stored to a scalar variable.
+    Scalar(Rc<CID<ServiceResultAggregate>>),
+    /// The call value is stored to a stream variable.
     Stream {
-        cid: Rc<CID<JValue>>,
+        cid: Rc<CID<ServiceResultAggregate>>,
         generation: u32,
     },
+    /// The call value is not stored.
+    Unused(Rc<CID<ServiceResultAggregate>>),
+}
+
+/// Please note that second argument is a JSON-serialized value, not a plain error message.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct CallServiceFailed(pub i32, pub Rc<String>);
+
+impl CallServiceFailed {
+    pub fn new(ret_code: i32, message: Rc<String>) -> Self {
+        Self(ret_code, message)
+    }
+
+    pub fn to_value(&self) -> JValue {
+        serde_json::to_value(self).expect("default serializer shouldn't fail")
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+/// A proof of service result execution result.
+pub struct ServiceResultAggregate {
+    pub value_cid: Rc<CID<JValue>>,
+    /// Hash of the call arguments.
+    pub argument_hash: Rc<str>,
```

Comment:
Why it's `str`, not byte slice?

Reply:
I use CID for the argument hash; also, string serializes better.